lms/djangoapps/sga_api/views.py

```diff
@@ -0,0 +1,293 @@
+        auto_enroll = data.get('auto_enroll', False)
+
+        if action not in ['enroll', 'unenroll']:
+            error_msg = "action must be enroll or unenroll"
+        elif not isinstance(courses, (list, tuple)):
+            error_msg = "courses must be a list of courses id "
+        elif not isinstance(users, (list, tuple)):
+            error_msg =  "users must be a list of identifiers(username or email)"
+        else:
+            error_msg = None
+
+        if error_msg:
+            return Response(
+                data={"error_message": error_msg},
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            result = self._enroll_unenrol_users_for_courses(
+                courses, users, request, action, email_students, auto_enroll
+            )
+
+            return Response(
+                data={
+                    'action': action,
+                    'courses': result,
+                    'email_students':email_students,
+                    'auto_enroll': auto_enroll,
+                },
+                status=status.HTTP_200_OK
+            )
+        except Exception as e:
+            log.error(e.message)
+            return Response(
+                data={"error_message": e.message},
+                status=status.HTTP_500_INTERNAL_SERVER_ERROR,
+            )
+
+    def _enroll_unenrol_users_for_courses(
+        self, list_of_courses_id, list_of_users_email,
+        request, action='enroll', email_students=False, auto_enroll=False
+    ):
+        courses = {}
+        for course_id in list_of_courses_id:
+            results = self._enroll_unenrol_users_for_course(
+                course_id, list_of_users_email, request,
+                action, email_students, auto_enroll
+            )
+            courses[course_id] = self._make_result_dict_for_course(
+                 results, action, auto_enroll
+            )
+        return courses
+
+    def _enroll_unenrol_users_for_course(
+            self, course_id, list_of_users_email, request,
+            action='enroll', email_students=False, auto_enroll=False
+    ):
+        result_list = []
+        for users_email in list_of_users_email:
+    
+            try:
+                user = get_student_from_identifier(users_email)
+                email = user.email
+                language = get_user_email_language(user)
+            except User.DoesNotExist:
+                if auto_enroll and action == 'enroll':
+                    email = users_email
+                    language = None
+                else:
+                    result_list.append(
+                        self._make_result_dict_for_identifier(
+                            users_email=users_email,
+                            error='User with {} email does not exist'.format(users_email),
+                            after={}, before={}
+                        )
+                    )
+                    continue
+            try:
```

Comment:
too many try - except blocks one would be fair enough.

Reply:
I use the first try-except block for check user. The user must be enabled if auto_enroll=false. But if auto_enroll=true user can absent. And in this case, the code can continue to work with email from the request.